tests/integration/utils.py

```diff
@@ -463,3 +472,165 @@ def _validate_expected_dep_file_contents(dep_contents_file: Path, output_dir: Pa
+    >>> affects_pm(Path('cachi2/core/package_managers/rpm/main.py'))
+    True
+    >>> affects_pm(Path('cachi2/core/package_managers/general.py'))
+    False
+    >>> affects_pm(Path('cachi2/core/package_managers/gomod.py'))
+    True
+    """
+
+    def name_belongs_to_a_pm(change: Path) -> bool:
+        return name_of(change) in SUPPORTED_PMS or change.parent.stem in SUPPORTED_PMS
+
+    def affects_pm_directly(change: Path) -> bool:
+        return "package_managers" in change.parts and name_belongs_to_a_pm(change)
+
+    def affects_pm_tests(change: Path) -> bool:
+        return tested_object_name(change) in SUPPORTED_PMS
+
+    return affects_pm_directly(change) or affects_pm_tests(change)
+
+
+def pm_name(pm_change: Path) -> str:
+    """Extract package manager name from a known package manager-related change.
+
+    >>> pm_name(Path('tests/integration/test_gomod.py'))
+    'gomod'
+    >>> pm_name(Path('cachi2/core/package_managers/rpm/main.py'))
+    'rpm'
+    >>> pm_name(Path('cachi2/core/package_managers/pip.py'))
+    'pip'
+    """
+    if (name := name_of(pm_change)) in SUPPORTED_PMS:
+        return name
+    elif (name := tested_object_name(pm_change)) in SUPPORTED_PMS:
+        return name
+    else:
+        return pm_change.parent.stem
+
+
+def affected_package_managers(pm_changes: tuple[Path, ...]) -> set[str]:
+    return set(pm_name(c) for c in pm_changes)
+
+
+def is_testable_code(c: Path) -> bool:
+    """Check if any actual code was affected by any of the changes.
+
+    Does this by checking if a change is in watched subtree.
+
+    >>> is_testable_code(Path('cachi2/core/config.py'))
+    True
+    >>> is_testable_code(Path('tests/integration/test_gomod.py'))
+    True
+    >>> is_testable_code(Path('tests/integration/utils.py'))
+    True
+    >>> is_testable_code(Path('tests/integration/conftest.py'))
+    True
+    >>> is_testable_code(Path('README.md'))
+    False
+    >>> is_testable_code(Path('reqruiements.txt'))
+    False
+    """
+    return any(c.is_relative_to(p) for p in PATHS_TO_CODE)
+
+
+def select_testable_changes(changes: tuple[Path, ...]) -> tuple[Path, ...]:
+    """Weed out changes that cannot be tested.
+
+    If a change is outside of paths to testable code it is dropped as if
+    it never happened. Any file or module outside of watched directories
+    will be rejected.
+    """
+    return tuple(c for c in changes if is_testable_code(c))
+
+
+def just_some_package_managers_were_affected_by(changes: tuple[Path, ...]) -> bool:
+    """Check that just package managers were affected.
+
+    If any code outside of package managers subtree was affected or if a module
+    shared by package managers was affected will return False.
+
+    >>> just_some_package_managers_were_affected_by((Path('tests/integration/test_pip.py'),))
+    True
+    >>> c = Path('cachi2/core/package_managers/pip.py'), Path('tests/integration/test_pip.py')
+    >>> just_some_package_managers_were_affected_by(c)
+    True
```

Comment:
FWIW I'm glad you added these to the docstring, really helpful.

Reply:
In case this is not a widely known fact one can change directory to tests/integration and then run `python3 -m doctest utils.py` and see which of these statements fail. 

I personally prefer doctests for utility functions because such functions are not busyness logic and should not be tested with UTs directly, but they do need some smoke-tests and I find short doctests to be a nearly perfect fit. Docstrings can be also easily folded away and not pollute the precious vertical space.

Reply:
TIL, thank you. It should be possible to do via pytest as well but `pytest --doctest-modules --doctest-continue-on-failure` is failing for me for some reason. Anyway, we should adopt this practice! I would have probably done something had I known about it :) .